### DISCUSSION TITLE
Mask plugin overwrites selected input characters when invalid characters are typed

### DISCUSSION THREAD

**Discussion by fuel-jengley** [alpinejs/alpine#3102] [7. Bugs (archived)]
Hi,

The current behavior of the mask plugin is to overwrite characters from the input if they are selected by the cursor and you type an invalid character. This could result in the invalid character not being filtered by the mask.

Shouldn't the mask ignore invalid characters regardless if any characters are selected or not?  

Two possible exceptions to this would be the backspace and delete keys. They of course are not filtered regardless if input characters are selected.

**Comment by kassah** [comment]
Here is a video demonstrating the issue when an 'a' character is pressed.

![Video demonstrating issue](https://user-images.githubusercontent.com/24927/184262642-36030a2c-866f-48ed-9083-bbeb35dd0cd5.mov)

**Comment by SimoTod** [comment]
I believe this is how it is supposed to work. the mask filtering happens on the input value so it needs to be updated first. It's the same with other famous mask plugins: see https://imask.js.org/

**Reply by kassah** [reply]
### Thoughts about Test Results
For built in browser fields that prevent invalid input most fields will prevent an invalid character from clearing selected text. There is one exception of Chrome input type number field will clear the value on an invalid character.

After surveying a few validators, it seems yes, for JS based mask libraries, this seems to be a common behavior. There are a few that handle it in the way the original poster desires, but many have strange behavior on selection or non-valid key press.

I'm hoping @calebporzio might chime in with his view on this.

### Built in Browser Masking Results
Testing validation on different browsers built in input limiters:

No -> Leaves selected value unchanged on a press of an invalid character press of "@" (i.e. shift 2)
Yes -> Will erase value selected on invalid character press of "@" (i.e. shift 2)
N/A -> Does not prevent the entry of invalid characters at all, validation happens later.

| Input Type | Chrome (104) | Firefox (103) |
| :---         |     :---:      |     :---:      |
| date   | No     | No    |
| datetime-local     | No       | No      |
| email*     | N/A       | N/A      |
| month     | No       | N/A      |
| number     | Yes       | N/A      |
| password     | N/A       | N/A      |
| search     | N/A       | N/A      |
| tel     | N/A       | N/A      |
| text     | N/A       | N/A      |
| time     | No       | No      |
| url     | N/A       | N/A      |
| week     | No       | N/A      |

\* I used a `!` instead of the usual test character, due to `@` being a valid input.

I used the following site for the tests: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/date

### JS Library Masking Results

No -> Leaves selected value unchanged on a press of an invalid character press of "@" (i.e. shift 2)
Yes -> Will erase value selected on invalid character press of "@" (i.e. shift 2)
N/A -> Does not prevent the entry of invalid characters at all, validation happens later.

| Library | Yes | No | Notes |
| :---  | :---:  | :---: | :--- |
| imask.js | X |  |  |
| cleave.js | X |  | The `@` replaces with a 5 in the Credit Card field? |
| Kendo UI [1] - Generic Masks |  | X | Regex rules seem to be 'No', However using built in '0' mask is 'Yes'  |
| Kendo UI [2] - Other Inputs |  | X  |  |
| Metric UI [3] | | X | Doesn't really allow selection of text, pressing shift adds 'Shift' to your string, very odd behavior |
| jQuery Mask Plugin [4] | X | | |

- [1] https://www.telerik.com/kendo-angular-ui/components/inputs/maskedtextbox/masks/
- [2] https://www.telerik.com/kendo-angular-ui/components/dateinputs/dateinput/date-ranges/
- [3] https://metroui.org.ua/input-mask.html
- [4] http://igorescobar.github.io/jQuery-Mask-Plugin/
